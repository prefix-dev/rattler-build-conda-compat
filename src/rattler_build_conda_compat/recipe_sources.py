from __future__ import annotations

import sys
import typing
from typing import Any, List, TypedDict, Union

from .conditional_list import ConditionalList, visit_conditional_list

if sys.version_info < (3, 11):
    from typing_extensions import NotRequired
else:
    from typing import NotRequired

if typing.TYPE_CHECKING:
    from collections.abc import Iterator, Mapping

OptionalUrlList = Union[str, List[str], None]


class Source(TypedDict):
    url: NotRequired[str]


def get_all_url_sources(recipe: Mapping[Any, Any]) -> Iterator[str]:
    """
    Get all url sources from the recipe. This can be from a list of sources,
    a single source, or conditional and its branches.

    Arguments
    ---------
    * `recipe` - The recipe to inspect. This should be a yaml object.

    Returns
    -------
    A list of sources.
    """

    sources = recipe.get("source", None)
    sources = typing.cast(ConditionalList[Source], sources)

    # Try getting all url top-level sources
    if sources is not None:
        source_list = visit_conditional_list(sources, None)
        for source in source_list:
            if url := source.get("url"):
                yield url

    outputs = recipe.get("outputs", None)
    if outputs is None:
        return

    outputs = visit_conditional_list(outputs, None)
    for output in outputs:
        sources = output.get("source", None)
        sources = typing.cast(ConditionalList[Source], sources)
        if sources is None:
            continue
        source_list = visit_conditional_list(sources, None)
        for source in source_list:
            if url := source.get("url"):
                yield url
