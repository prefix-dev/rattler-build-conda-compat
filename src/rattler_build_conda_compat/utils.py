import fnmatch
import itertools
from logging import getLogger
import os
from pathlib import Path
from typing import Any, Dict, Iterable


VALID_METAS = ("recipe.yaml",)


def islist(arg, uniform=False, include_dict=True):
    """
    Check whether `arg` is a `list`. Optionally determine whether the list elements
    are all uniform.

    When checking for generic uniformity (`uniform=True`) we check to see if all
    elements are of the first element's type (`type(arg[0]) == type(arg[1])`). For
    any other kinds of uniformity checks are desired provide a uniformity function:

    .. code-block:: pycon
        # uniformity function checking if elements are str and not empty
        >>> truthy_str = lambda e: isinstance(e, str) and e
        >>> islist(["foo", "bar"], uniform=truthy_str)
        True
        >>> islist(["", "bar"], uniform=truthy_str)
        False
        >>> islist([0, "bar"], uniform=truthy_str)
        False

    .. note::
        Testing for uniformity will consume generators.

    :param arg: Object to ensure is a `list`
    :type arg: any
    :param uniform: Whether to check for uniform or uniformity function
    :type uniform: bool, function, optional
    :param include_dict: Whether to treat `dict` as a `list`
    :type include_dict: bool, optional
    :return: Whether `arg` is a `list`
    :rtype: bool
    """
    if isinstance(arg, str) or not isinstance(arg, Iterable):
        # str and non-iterables are not lists
        return False
    elif not include_dict and isinstance(arg, dict):
        # do not treat dict as a list
        return False
    elif not uniform:
        # short circuit for non-uniformity
        return True

    # NOTE: not checking for Falsy arg since arg may be a generator
    # WARNING: if uniform != False and arg is a generator then arg will be consumed

    if uniform is True:
        arg = iter(arg)
        try:
            etype = type(next(arg))
        except StopIteration:
            # StopIteration: list is empty, an empty list is still uniform
            return True
        # check for explicit type match, do not allow the ambiguity of isinstance
        uniform = lambda e: type(e) == etype  # noqa: E731

    try:
        return all(uniform(e) for e in arg)
    except (ValueError, TypeError):
        # ValueError, TypeError: uniform function failed
        return False


def ensure_list(arg, include_dict=True):
    """
    Ensure the object is a list. If not return it in a list.

    :param arg: Object to ensure is a list
    :type arg: any
    :param include_dict: Whether to treat `dict` as a `list`
    :type include_dict: bool, optional
    :return: `arg` as a `list`
    :rtype: list
    """
    if arg is None:
        return []
    elif islist(arg, include_dict=include_dict):
        return list(arg)
    else:
        return [arg]


def rec_glob(path, patterns, ignores=None):
    """
    Recursively searches path for filename patterns.

    :param path: path within to search for files
    :param patterns: list of filename patterns to search for
    :param ignore: list of directory patterns to ignore in search
    :return: list of paths in path satisfying patterns/ignore
    """
    patterns = ensure_list(patterns)
    ignores = ensure_list(ignores)

    for path, dirs, files in os.walk(path):
        # remove directories to ignore
        for ignore in ignores:
            for d in fnmatch.filter(dirs, ignore):
                dirs.remove(d)

        # return filepaths that match a pattern
        for pattern in patterns:
            for f in fnmatch.filter(files, pattern):
                yield os.path.join(path, f)


def find_recipe(path):
    """
    vendored from conda_build.utils to persist same API flow

    recurse through a folder, locating valid meta files (see VALID_METAS).  Raises error if more than one is found.

    Returns full path to meta file to be built.

    If we have a base level meta file and other supplemental (nested) ones, use the base level.
    """
    # if initial path is absolute then any path we find (via rec_glob)
    # will also be absolute
    if not os.path.isabs(path):
        path = os.path.normpath(os.path.join(os.getcwd(), path))

    if os.path.isfile(path):
        if os.path.basename(path) in VALID_METAS:
            return path
        raise OSError("{} is not a valid meta file ({})".format(path, ", ".join(VALID_METAS)))

    results = list(rec_glob(path, VALID_METAS, ignores=(".AppleDouble",)))

    if not results:
        raise OSError("No meta files ({}) found in {}".format(", ".join(VALID_METAS), path))

    if len(results) == 1:
        return results[0]

    # got multiple valid meta files
    # check if a meta file is defined on the base level in which case use that one

    metas = [m for m in VALID_METAS if os.path.isfile(os.path.join(path, m))]
    if len(metas) == 1:
        getLogger(__name__).warn(
            "Multiple meta files found. "
            f"The {metas[0]} file in the base directory ({path}) "
            "will be used."
        )
        return os.path.join(path, metas[0])

    raise OSError("More than one meta files ({}) found in {}".format(", ".join(VALID_METAS), path))


def has_recipe(recipe_dir: Path) -> bool:
    """

    verify if recipe_dir contains recipe.yaml

    it is used to detect whetever old or recipe should be used


    """
    try:
        recipe_path = find_recipe(recipe_dir)
        if recipe_path:
            return True
        return False
    except OSError:
        return False


def remove_empty_keys(some_dict: Dict[str, Any]) -> Dict[str, Any]:
    filtered_dict = {}
    for key, value in some_dict.items():
        if isinstance(value, list) and len(value) == 0:
            continue
        filtered_dict[key] = value

    return filtered_dict


def flatten_lists(some_dict: Dict[str, Any]) -> Dict[str, Any]:
    result_dict: Dict[str, Any] = {}
    for key, value in some_dict.items():
        if isinstance(value, dict):
            result_dict[key] = flatten_lists(value)
        elif isinstance(value, list) and value and isinstance(value[0], list):
            result_dict[key] = list(itertools.chain(*value))
        else:
            result_dict[key] = value

    return result_dict
