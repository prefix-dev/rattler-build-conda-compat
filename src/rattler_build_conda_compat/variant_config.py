
from itertools import product
from typing import List, Mapping

def variant_combinations(data) -> List[Mapping[str, str]]:
    zip_keys = data.pop("zip_keys", [])
    # Separate the keys that need to be zipped from the rest
    zip_keys_flat = [item for sublist in zip_keys for item in sublist]
    other_keys = [key for key in data.keys() if key not in zip_keys_flat]

    # Create combinations for non-zipped keys
    other_combinations = list(product(*[data[key] for key in other_keys]))

    # Create zipped combinations
    zipped_combinations = [list(zip(*[data[key] for key in zip_group])) for zip_group in zip_keys]

    # Combine zipped combinations
    zipped_product = list(product(*zipped_combinations))

    # Combine all results into dictionaries
    final_combinations = []
    for other_combo in other_combinations:
        for zipped_combo in zipped_product:
            combined = {}
            # Add non-zipped items
            for key, value in zip(other_keys, other_combo):
                combined[key] = str(value)
            # Add zipped items
            for zip_group, zip_values in zip(zip_keys, zipped_combo):
                for key, value in zip(zip_group, zip_values):
                    combined[key] = str(value)
            final_combinations.append(combined)

    return final_combinations

