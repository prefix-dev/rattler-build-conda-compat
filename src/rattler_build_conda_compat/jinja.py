from __future__ import annotations

from typing import Any, TypedDict

import jinja2
import yaml
from jinja2 import DebugUndefined

from rattler_build_conda_compat.loader import load_yaml


class RecipeWithContext(TypedDict, total=False):
    context: dict[str, str]


class _MissingUndefined(DebugUndefined):
    def __str__(self) -> str:
        """
        By default, `DebugUndefined` return values in the form `{{ value }}`.
        `rattler-build` has a different syntax, so we need to override this method,
        and return the value in the form `${{ value }}`.
        """
        return f"${super().__str__()}"


def jinja_env() -> jinja2.Environment:
    """
    Create a `rattler-build` specific Jinja2 environment with modified syntax.
    """
    return jinja2.Environment(
        variable_start_string="${{",
        variable_end_string="}}",
        trim_blocks=True,
        lstrip_blocks=True,
        autoescape=True,
        undefined=_MissingUndefined,
    )


def load_recipe_context(context: dict[str, str], jinja_env: jinja2.Environment) -> dict[str, str]:
    """
    Load all string values from the context dictionary as Jinja2 templates.
    """
    # Process each key-value pair in the dictionary
    for key, value in context.items():
        # If the value is a string, render it as a template
        if isinstance(value, str):
            template = jinja_env.from_string(value)
            rendered_value = template.render(context)
            context[key] = rendered_value

    return context


def render_recipe_with_context(recipe_content: RecipeWithContext) -> dict[str, Any]:
    """
    Render the recipe using known values from context section.
    Unknown values are not evaluated and are kept as it is.

    Examples:
    ---
    ```python
    >>> from pathlib import Path
    >>> from rattler_build_conda_compat.loader import load_yaml
    >>> recipe_content = load_yaml((Path().resolve() / "tests" / "data" / "eval_recipe_using_context.yaml").read_text())
    >>> evaluated_context = render_recipe_with_context(recipe_content)
    >>> assert "my_value-${{ not_present_value }}" == evaluated_context["build"]["string"]
    >>>
    ```
    """
    env = jinja_env()
    context = recipe_content.get("context", {})
    # load all context templates
    context_templates = load_recipe_context(context, env)

    # render the rest of the document with the values from the context
    # and keep undefined expressions _as is_.
    template = env.from_string(yaml.dump(recipe_content))
    rendered_content = template.render(context_templates)
    return load_yaml(rendered_content)
