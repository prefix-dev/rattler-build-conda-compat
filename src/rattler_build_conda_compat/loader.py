from __future__ import annotations

from contextlib import contextmanager
from typing import TYPE_CHECKING, Any

import yaml

from rattler_build_conda_compat.conditional_list import visit_conditional_list
from rattler_build_conda_compat.utils import flatten_lists, remove_empty_keys

if TYPE_CHECKING:
    from collections.abc import Iterator
    from os import PathLike

SELECTOR_OPERATORS = ("and", "or", "not")


class RecipeLoader(yaml.BaseLoader):
    _namespace: dict[str, Any] | None = None
    _allow_missing_selector: bool = False

    @classmethod
    @contextmanager
    def with_namespace(
        cls: type[RecipeLoader],
        namespace: dict[str, Any] | None,
        *,
        allow_missing_selector: bool = False,
    ) -> Iterator[None]:
        try:
            cls._namespace = namespace
            cls._allow_missing_selector = allow_missing_selector
            yield
        finally:
            del cls._namespace

    def construct_sequence(  # noqa: C901, PLR0912
        self,
        node: yaml.ScalarNode | yaml.SequenceNode | yaml.MappingNode,
        deep: bool = False,  # noqa: FBT002, FBT001,
    ) -> list[yaml.ScalarNode]:
        """deep is True when creating an object/mapping recursively,
        in that case want the underlying elements available during construction
        """
        # find if then else selectors
        for sequence_idx, child_node in enumerate(node.value[:]):
            # if then is only present in MappingNode

            if isinstance(child_node, yaml.MappingNode):
                # iterate to find if there is IF first

                the_evaluated_one = None
                for idx, (key_node, value_node) in enumerate(child_node.value):
                    if key_node.value == "if":
                        # we catch the first one, let's try to find next pair of (then | else)
                        then_node_key, then_node_value = child_node.value[idx + 1]

                        if then_node_key.value != "then":
                            msg = "cannot have if without then, please reformat your variant file"
                            raise ValueError(msg)

                        try:
                            _, else_node_value = child_node.value[idx + 2]
                        except IndexError:
                            _, else_node_value = None, None

                        to_be_eval = f"{value_node.value}"

                        if self._allow_missing_selector:
                            split_selectors = [
                                selector
                                for selector in to_be_eval.split()
                                if selector not in SELECTOR_OPERATORS
                            ]
                            for selector in split_selectors:
                                if self._namespace and selector not in self._namespace:
                                    cleaned_selector = selector.strip("(").rstrip(")")
                                    self._namespace[cleaned_selector] = True

                        evaled = eval(to_be_eval, self._namespace)  # noqa: S307
                        if evaled:
                            the_evaluated_one = then_node_value
                        elif else_node_value:
                            the_evaluated_one = else_node_value

                        if the_evaluated_one:
                            node.value.remove(child_node)
                            node.value.insert(sequence_idx, the_evaluated_one)
                        else:
                            # neither the evaluation or else node is present, so we remove this if
                            node.value.remove(child_node)

        if not isinstance(node, yaml.SequenceNode):
            raise TypeError(
                None,
                None,
                f"expected a sequence node, but found {node.id!s}",
                node.start_mark,
            )

        return [self.construct_object(child, deep=deep) for child in node.value]


def load_yaml(content: str | bytes) -> Any:  # noqa: ANN401
    return yaml.load(content, Loader=yaml.BaseLoader)  # noqa: S506


def parse_recipe_config_file(
    path: PathLike[str], namespace: dict[str, Any] | None, *, allow_missing_selector: bool = False
) -> dict[str, Any]:
    with open(path) as f, RecipeLoader.with_namespace(
        namespace, allow_missing_selector=allow_missing_selector
    ):
        content = yaml.load(f, Loader=RecipeLoader)  # noqa: S506
    return flatten_lists(remove_empty_keys(content))


def load_all_requirements(content: dict[str, Any]) -> dict[str, Any]:
    requirements_section = dict(content.get("requirements", {}))
    if not requirements_section:
        return {}

    for section in requirements_section:
        section_reqs = requirements_section[section]
        if not section_reqs:
            continue

        requirements_section[section] = list(visit_conditional_list(section_reqs))

    return requirements_section
