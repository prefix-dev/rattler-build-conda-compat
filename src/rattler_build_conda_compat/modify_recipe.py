from __future__ import annotations

import copy
import hashlib
import io
import logging
import re
from typing import TYPE_CHECKING, Any, Literal

import requests
from ruamel.yaml import YAML

from rattler_build_conda_compat.jinja.jinja import jinja_env, load_recipe_context
from rattler_build_conda_compat.recipe_sources import Source, get_all_sources

if TYPE_CHECKING:
    from pathlib import Path

logger = logging.getLogger(__name__)

yaml = YAML()
yaml.preserve_quotes = True
yaml.width = 4096
yaml.indent(mapping=2, sequence=4, offset=2)


def _update_build_number_in_context(recipe: dict[str, Any], new_build_number: int) -> bool:
    for key in recipe.get("context", {}):
        if key.startswith("build_") or key == "build":
            recipe["context"][key] = new_build_number
            return True
    return False


def _update_build_number_in_recipe(recipe: dict[str, Any], new_build_number: int) -> bool:
    is_modified = False
    if "build" in recipe and "number" in recipe["build"]:
        recipe["build"]["number"] = new_build_number
        is_modified = True

    if "outputs" in recipe:
        for output in recipe["outputs"]:
            if "build" in output and "number" in output["build"]:
                output["build"]["number"] = new_build_number
                is_modified = True

    return is_modified


def update_build_number(file: Path, new_build_number: int = 0) -> str:
    """
    Update the build number in the recipe file.

    Arguments:
    ----------
    * `file` - The path to the recipe file.
    * `new_build_number` - The new build number to use. (default: 0)

    Returns:
    --------
    The updated recipe as a string.
    """
    with file.open("r") as f:
        data = yaml.load(f)
    build_number_modified = _update_build_number_in_context(data, new_build_number)
    if not build_number_modified:
        _update_build_number_in_recipe(data, new_build_number)

    with io.StringIO() as f:
        yaml.dump(data, f)
        return f.getvalue()


class CouldNotUpdateVersionError(Exception):
    NO_CONTEXT = "Could not find context in recipe"
    NO_VERSION = "Could not find version in recipe context"

    def __init__(self, message: str = "Could not update version") -> None:
        self.message = message
        super().__init__(self.message)


class Hash:
    def __init__(self, hash_type: Literal["md5", "sha256"], hash_value: str) -> None:
        self.hash_type = hash_type
        self.hash_value = hash_value

    def __str__(self) -> str:
        return f"{self.hash_type}: {self.hash_value}"


def _has_jinja_version(url: str) -> bool:
    """Check if the URL has a jinja `${{ version }}` in it."""
    pattern = r"\${{\s*version"
    return re.search(pattern, url) is not None


def update_hash(source: Source, url: str, hash_: Hash | None) -> None:
    """
    Update the sha256 hash in the source dictionary.

    Arguments:
    ----------
    * `source` - The source dictionary to update.
    * `url` - The URL to download and hash (if no hash is provided).
    * `hash_` - The hash to use. If not provided, the file will be downloaded and `sha256` hashed.
    """
    hash_type = hash_.hash_type if hash_ is not None else "sha256"
    # delete all old hashes that we are not updating
    all_hash_types = {"md5", "sha256"}
    for key in all_hash_types - {hash_type}:
        if key in source:
            del source[key]

    if hash_ is not None:
        source[hash_.hash_type] = hash_.hash_value
    else:
        # download and hash the file
        hasher = hashlib.sha256()
        print(f"Retrieving and hashing {url}")
        with requests.get(url, stream=True, timeout=100) as r:
            for chunk in r.iter_content(chunk_size=4096):
                hasher.update(chunk)
        source["sha256"] = hasher.hexdigest()


def update_version(file: Path, new_version: str, hash_: Hash | None) -> str:
    """
    Update the version in the recipe file.

    Arguments:
    ----------
    * `file` - The path to the recipe file.
    * `new_version` - The new version to use.
    * `hash_type` - The hash type to use. If not provided, the file will be downloaded and `sha256` hashed.

    Returns:
    --------
    The updated recipe as a string.
    """

    with file.open("r") as f:
        data = yaml.load(f)

    if "context" not in data:
        raise CouldNotUpdateVersionError(CouldNotUpdateVersionError.NO_CONTEXT)
    if "version" not in data["context"]:
        raise CouldNotUpdateVersionError(CouldNotUpdateVersionError.NO_VERSION)

    data["context"]["version"] = new_version

    # set up the jinja context
    env = jinja_env()
    context = copy.deepcopy(data.get("context", {}))
    context_variables = load_recipe_context(context, env)
    # for r-recipes we add the default `cran_mirror` variable
    context_variables["cran_mirror"] = "https://cran.r-project.org"

    for source in get_all_sources(data):
        # render the whole URL and find the hash
        if "url" not in source:
            continue

        url = source["url"]
        if isinstance(url, list):
            url = url[0]

        if not _has_jinja_version(url):
            continue

        template = env.from_string(url)
        rendered_url = template.render(context_variables)

        update_hash(source, rendered_url, hash_)

    with io.StringIO() as f:
        yaml.dump(data, f)
        return f.getvalue()
