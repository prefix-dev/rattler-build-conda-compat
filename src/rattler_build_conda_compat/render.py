# mypy: ignore-errors

from __future__ import annotations

from collections import OrderedDict
import json
import os
from pathlib import Path
import subprocess
import sys
import tempfile
from typing import Any, Dict, List, Optional

from conda_build.metadata import (
    MetaData as CondaMetaData,
    OPTIONALLY_ITERABLE_FIELDS,
)
from conda_build.config import get_or_merge_config
from conda_build.variants import (
    filter_combined_spec_to_used_keys,
    get_default_variant,
    validate_spec,
    combine_specs,
)
from conda_build.metadata import get_selectors, check_bad_chrs
from conda_build.config import Config

from rattler_build_conda_compat.jinja.jinja import render_recipe_with_context
from rattler_build_conda_compat.loader import load_yaml, parse_recipe_config_file
from rattler_build_conda_compat.utils import _get_recipe_metadata, find_recipe
from rattler_build_conda_compat.yaml import _yaml_object


class MetaData(CondaMetaData):
    def __init__(
        self,
        path,
        rendered_recipe: Optional[dict] = None,
        config=None,
        variant=None,
    ):
        self.config: Config = get_or_merge_config(config, variant=variant)
        if os.path.isfile(path):
            self._meta_path = path
            self._meta_name = os.path.basename(path)
            self.path = os.path.dirname(path)
        else:
            self._meta_name = "recipe.yaml"
            self._meta_path = find_recipe(path)
            self.path = os.path.dirname(self._meta_path)

        self._rendered = False

        if not rendered_recipe:
            self.meta = self.parse_recipe()
            self.meta["about"] = self.meta.get("about", {})
            self.meta["extra"] = self.meta.get("extra", {})
        else:
            self.meta = rendered_recipe
            self._rendered = True
            self.meta["about"] = self.meta["recipe"].get("about", {})
            self.meta["extra"] = self.meta["recipe"].get("extra", {})

        self.final = True
        self.undefined_jinja_vars = []

        self.requirements_path = os.path.join(self.path, "requirements.txt")

    def parse_recipe(self) -> dict[str, Any]:
        recipe_path: Path = Path(self.path) / self._meta_name

        yaml_content = load_yaml(recipe_path.read_text())

        return render_recipe_with_context(yaml_content)

    def name(self) -> str:
        """
        Overrides the conda_build.metadata.MetaData.name method.
        Returns the name of the package.
        If recipe has multiple outputs, it will return the name of the `recipe` field.
        Otherwise it will return the name of the `package` field.

        Raises:
            - CondaBuildUserError: If the `name` contains bad characters.
            - ValueError: If the name is not lowercase or missing.

        """
        name = _get_recipe_metadata(self.meta, "name", rendered=self._rendered)

        if not name:
            raise ValueError(f"Error: package/name missing in: {self.meta_path!r}")

        if name != name.lower():
            raise ValueError(f"Error: package/name must be lowercase, got: {name!r}")

        check_bad_chrs(name, "package/name")
        return name

    def version(self) -> str:
        """
        Overrides the conda_build.metadata.MetaData.version method.
        Returns the version of the package.
        If recipe has multiple outputs, it will return the version of the `recipe` field.
        Otherwise it will return the version of the `package` field.

        Raises:
            - CondaBuildUserError: If the `version` contains bad characters.
            - ValueError: If the version starts with a period or version is missing.
        """
        version = _get_recipe_metadata(self.meta, "version", rendered=self._rendered)

        if not version:
            raise ValueError(f"Error: package/version missing in: {self.meta_path!r}")

        check_bad_chrs(version, "package/version")
        if version.startswith("."):
            raise ValueError(f"Fully-rendered version can't start with period -  got {version!r}")
        return version

    def render_recipes(self, variants) -> List[Dict]:
        build_platform_and_arch = f"{self.config.platform}-{self.config.arch}"
        target_platform_and_arch = f"{self.config.host_platform}-{self.config.host_arch}"

        print("***************************************************")

        print(f"build_platform_and_arch: {build_platform_and_arch}")
        print(f"target_platform_and_arch: {target_platform_and_arch}")
        print(f"variants: {variants}")
        print(f"config: {self.config}")

        print("***************************************************")

        yaml = _yaml_object()
        try:
            with tempfile.NamedTemporaryFile(mode="w+") as outfile:
                with tempfile.NamedTemporaryFile(mode="w") as variants_file:
                    # dump variants in our variants that will be used to generate recipe
                    if variants:
                        yaml.dump(variants, variants_file)

                    variants_path = variants_file.name

                    run_args = [
                        "rattler-build",
                        "build",
                        "--render-only",
                        "--recipe",
                        self.path,
                        "--target-platform",
                        target_platform_and_arch,
                        "--build-platform",
                        build_platform_and_arch,
                    ]

                    if variants:
                        run_args.extend(["-m", variants_path])
                    subprocess.run(run_args, check=True, stdout=outfile, env=os.environ)

                    outfile.seek(0)
                    content = outfile.read()
                    metadata = json.loads(content)
            return metadata if isinstance(metadata, list) else [metadata]

        except Exception as e:
            raise e

    def get_used_vars(self, force_top_level=False, force_global=False):
        if "build_configuration" not in self.meta:
            # it could be that we skip build for this platform
            # so no variants have been discovered
            # return empty
            return set()

        used_vars = [
            var.replace("-", "_") for var in self.meta["build_configuration"]["variant"].keys()
        ]

        # in conda-build target-platform is not returned as part of yaml vars
        # so it's included manually
        # in our case it is always present in build_configuration.variant
        # so we remove it when it's noarch
        if "target_platform" in self.config.variant and self.noarch:
            used_vars.remove("target_platform")

        return set(used_vars)

    def get_used_variant(self) -> Dict:
        if "build_configuration" not in self.meta:
            # it could be that we skip build for this platform
            # so no variants have been discovered
            # return empty
            return {}

        used_variant = dict(self.meta["build_configuration"]["variant"])

        used_variant_key_normalized = {}

        for key, value in used_variant.items():
            normalized_key = key.replace("-", "_")
            used_variant_key_normalized[normalized_key] = value

        # in conda-build target-platform is not returned as part of yaml vars
        # so it's included manually
        # in our case it is always present in build_configuration.variant
        # so we remove it when it's noarch
        if "target_platform" in used_variant_key_normalized and self.noarch:
            used_variant_key_normalized.pop("target_platform")

        return used_variant_key_normalized

    def get_used_loop_vars(self, force_top_level=False, force_global=False):
        return {
            var
            for var in self.get_used_vars(
                force_top_level=force_top_level, force_global=force_global
            )
            if var in self.get_loop_vars()
        }

    def get_section(self, name):
        if not self._rendered:
            section = self.meta.get(name)
        else:
            section = self.meta.get("recipe", {}).get(name)

        if name in OPTIONALLY_ITERABLE_FIELDS:
            if not section:
                return []
            elif isinstance(section, dict):
                return [section]
            elif not isinstance(section, list):
                raise ValueError(f"Expected {name} to be a list")
        else:
            if not section:
                return {}
            elif not isinstance(section, dict):
                raise ValueError(f"Expected {name} to be a dict")

        return section


def render_recipe(
    recipe_path,
    config=None,
    variants=None,
) -> List[MetaData]:
    """Returns a list of tuples, each consisting of

    (metadata-object, needs_download, needs_render_in_env)

    You get one tuple per variant.  Outputs are not factored in here (subpackages won't affect these
    results returned here.)
    """

    metadata = MetaData(recipe_path, config=config)
    recipes = metadata.render_recipes(variants)

    metadatas: list[MetaData] = []
    if not recipes:
        return [metadata]

    for recipe in recipes:
        metadata = MetaData(recipe_path, rendered_recipe=recipe, config=config)
        # just to have the same interface as conda_build
        metadatas.append(metadata)

    return metadatas


def render(
    recipe_path: os.PathLike,
    config: Optional[Config] = None,
    variants: Optional[Dict[str, Any] | None] = None,
    **kwargs,
):
    """Given path to a recipe, return the MetaData object(s) representing that recipe, with jinja2
       templates evaluated.

    Returns a list of (metadata, needs_download, needs_reparse in env) tuples
    """

    config = get_or_merge_config(config, **kwargs)

    arg = recipe_path
    if os.path.isfile(arg):
        if arg.endswith(".yaml"):
            recipe_dir = os.path.dirname(arg)
        else:
            raise ValueError("Recipe don't have a valid extension: %s" % arg)
    else:
        recipe_dir = os.path.abspath(arg)

    metadata_tuples = render_recipe(
        recipe_dir,
        config=config,
        variants=variants,
    )

    for m in metadata_tuples:
        if not hasattr(m.config, "variants") or not m.config.variant:
            m.config.ignore_system_variants = True

            if os.path.isfile(os.path.join(m.path, "variants.yaml")):
                m.config.variant_config_files = [os.path.join(m.path, "variants.yaml")]

            used_variant = m.get_used_variant()

            package_variants = rattler_get_package_variants(m, variants=variants)

            m.config.variants = package_variants[:]

            # we need to discard variants that we don't use
            for pkg_variant in package_variants[:]:
                for (
                    used_variant_key,
                    used_variant_value,
                ) in used_variant.items():
                    if used_variant_key in pkg_variant:
                        if (
                            pkg_variant[used_variant_key] != used_variant_value
                            and pkg_variant in package_variants
                        ):
                            package_variants.remove(pkg_variant)

            m.config.variant = package_variants[0]

            # These are always the full set.  just 'variants' is the one that gets
            #     used mostly, and can be reduced
            m.config.input_variants = m.config.variants
            m.config.variants = package_variants

    return [(m, False, False) for m in metadata_tuples]


def get_package_combined_spec(recipedir_or_metadata, config, variants=None):
    # this function is *vendored* version of
    # get_package_combined_spec from conda_build
    # with few changes to support rattler-build

    config = recipedir_or_metadata.config
    namespace = get_selectors(config)
    variants_paths = config.variant_config_files

    specs = OrderedDict(internal_defaults=get_default_variant(config))

    for variant_path in variants_paths:
        specs[variant_path] = parse_recipe_config_file(variant_path, namespace)

    # this is the override of the variants from files and args with values from CLI or env vars
    if hasattr(config, "variant") and config.variant:
        specs["config.variant"] = config.variant
    if variants:
        specs["argument_variants"] = variants

    for f, spec in specs.items():
        validate_spec(f, spec)

    # this merges each of the specs, providing a debug message when a given setting is overridden
    #      by a later spec
    combined_spec = combine_specs(specs, log_output=config.verbose)

    return combined_spec, specs


def rattler_get_package_variants(recipedir_or_metadata, config=None, variants=None):
    # this function is *vendored* version of
    # get_package_variants from conda_build
    # with few changes to support rattler-build
    combined_spec, specs = get_package_combined_spec(
        recipedir_or_metadata, config=config, variants=variants
    )
    return filter_combined_spec_to_used_keys(combined_spec, specs=specs)
