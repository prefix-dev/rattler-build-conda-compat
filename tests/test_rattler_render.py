from __future__ import annotations

import os
from typing import TYPE_CHECKING, Any
from pathlib import Path

from rattler_build_conda_compat.loader import parse_recipe_config_file, load_yaml, eval_jinja
from rattler_build_conda_compat.render import render

if TYPE_CHECKING:
    from pathlib import Path


def test_render_recipe(python_recipe: Path, unix_namespace: dict[str, Any], snapshot) -> None:
    variants = parse_recipe_config_file(str(python_recipe / "variants.yaml"), unix_namespace)

    rendered = render(str(python_recipe), variants=variants, platform="linux", arch="64")

    all_used_variants = [meta[0].get_used_variant() for meta in rendered]

    assert len(all_used_variants) == 2

    assert snapshot == all_used_variants


def test_environ_is_passed_to_rattler_build(env_recipe, snapshot) -> None:
    try:
        os.environ["TEST_SHOULD_BE_PASSED"] = "false"
        rendered = render(str(env_recipe), platform="linux", arch="64")
        all_used_variants = [meta[0].meta for meta in rendered]
        assert len(all_used_variants) == 1
        # for this scenario recipe should not be rendered
        assert snapshot == all_used_variants

        os.environ["TEST_SHOULD_BE_PASSED"] = "true"

        rendered = render(str(env_recipe), platform="linux", arch="64")

        all_used_variants = [meta[0].meta for meta in rendered]
        assert len(all_used_variants) == 1
        # for this scenario recipe should be rendered
        assert snapshot == all_used_variants[0]["build_configuration"]["variant"]

    finally:
        os.environ.pop("TEST_SHOULD_BE_PASSED", None)


def test_render_context(snapshot) -> None:
    recipe = Path("tests/data/context.yaml")
    with recipe.open() as f:
        recipe_yaml = load_yaml(f)

    rendered = eval_jinja(recipe_yaml)
    assert rendered == snapshot