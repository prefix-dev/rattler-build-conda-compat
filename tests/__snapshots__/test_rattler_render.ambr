# serializer version: 1
# name: test_environ_is_passed_to_rattler_build
  list([
    CommentedMap({
      'package': CommentedMap({
        'name': 'py-test',
        'version': '1.0.0',
      }),
      'build': CommentedMap({
        'skip': CommentedSeq([
          "env.get('TEST_SHOULD_BE_PASSED') == 'false'",
        ]),
      }),
      'requirements': CommentedMap({
        'build': CommentedSeq([
          "${{ compiler('c') }}",
          "${{ compiler('cuda') }}",
        ]),
        'host': CommentedSeq([
          'python',
        ]),
        'run': CommentedSeq([
          'python',
        ]),
      }),
      'about': dict({
      }),
      'extra': dict({
        'final': True,
      }),
    }),
  ])
# ---
# name: test_environ_is_passed_to_rattler_build.1
  dict({
    'target_platform': 'linux-64',
  })
# ---
# name: test_render_recipe
  list([
    dict({
      'python': '3.8.* *_cpython',
      'target_platform': 'linux-64',
    }),
    dict({
      'python': '3.9.* *_cpython',
      'target_platform': 'linux-64',
    }),
  ])
# ---
